src/coreComponents/linearAlgebra/interfaces/hypre/HypreExport.cpp

```diff
@@ -0,0 +1,195 @@
+  : m_targetRank( targetRank )
+{
+  // make a sub-communicator for scatter and ensure target rank is mapped to 0 in new comm
+  int const rank = MpiWrapper::commRank( mat.getComm() );
+  int const color = ( mat.numLocalRows() > 0 ) ? 0 : MPI_UNDEFINED;
+  int const key = ( rank == m_targetRank ) ? 0 : ( rank < m_targetRank ) ? rank + 1 : rank;
+  m_subComm = MpiWrapper::commSplit( mat.getComm(), color, key );
+}
+
+HypreExport::~HypreExport()
+{
+  if( m_subComm != MPI_COMM_NULL )
+  {
+    MpiWrapper::commFree( m_subComm );
+  }
+}
+
+template< typename OFFSET_TYPE, typename COLUMN_TYPE >
+void HypreExport::exportCRS( HypreMatrix const & mat,
+                             OFFSET_TYPE * const rowOffsets,
+                             COLUMN_TYPE * const colIndices,
+                             real64 * const values ) const
+{
+  int const rank = MpiWrapper::commRank( mat.getComm() );
+
+  // import on target rank if needed, or extract diag+offdiag part in each rank
+  hypre_CSRMatrix * localMatrix;
+  if( m_targetRank < 0 )
+  {
+    localMatrix = hypre_MergeDiagAndOffd( mat.unwrapped() );
+  }
+  else
+  {
+    GEOSX_ERROR_IF( rank == m_targetRank && !( rowOffsets && colIndices && values ),
+                    "HypreExport: must pass non-null pointers on the target rank" );
+    localMatrix = hypre_ParCSRMatrixToCSRMatrixAll( mat.unwrapped() );
+    GEOSX_ERROR_IF( rank == m_targetRank && !localMatrix, "HypreExport: matrix is empty on target rank" );
+  }
+
+  // export the raw CRS data
+  if( m_targetRank < 0 || m_targetRank == rank )
+  {
+    HYPRE_Int const numRows = hypre_CSRMatrixNumRows( localMatrix );
+    HYPRE_Int const numNz = hypre_CSRMatrixNumNonzeros( localMatrix );
+    HYPRE_Int const * const ia = hypre_CSRMatrixI( localMatrix );
+    HYPRE_Real const * const va = hypre_CSRMatrixData( localMatrix );
+
+    std::transform( ia, ia + numRows + 1, rowOffsets, LvArray::integerConversion< OFFSET_TYPE, HYPRE_Int > );
+    std::copy( va, va + numNz, values );
+
+    // WTF: We have to handle two cases differently because hypre uses two different struct members (!)
+    //      to store the column indices depending on how we obtained the local matrix (this is fucked).
+    if( m_targetRank < 0 )
+    {
+      HYPRE_BigInt const * const ja = hypre_CSRMatrixBigJ( localMatrix );
+      std::transform( ja, ja + numNz, colIndices, LvArray::integerConversion< COLUMN_TYPE, HYPRE_BigInt > );
+    }
+    else
+    {
+      HYPRE_Int const * const ja = hypre_CSRMatrixJ( localMatrix );
+      std::transform( ja, ja + numNz, colIndices, LvArray::integerConversion< COLUMN_TYPE, HYPRE_Int > );
+
+      // Sort the values by column index after copying (some solvers expect this, but not guaranteed)
+      for( localIndex i = 0; i < numRows; ++i )
+      {
+        using LvArray::sortedArrayManipulation::dualSort;
+        dualSort( colIndices + rowOffsets[i], colIndices + rowOffsets[i + 1], values + rowOffsets[i] );
+      }
```

Comment:
Good idea. Currently the parallel export is only used by SuperLU_Dist which does not care, but since the export functionality is now isolated, it's probably if the behavior is consistent between both cases.